lib/chef/resource.rb

```diff
@@ -871,23 +864,197 @@ def after_created
+      #
+      # Resources in the action recipe may notify and subscribe to other resources
+      # within the action recipe, but cannot notify or subscribe to resources
+      # in the main Chef run.
+      #
+      # Resource actions are *inheritable*: if resource A defines `action :create`
+      # and B is a subclass of A, B gets all of A's actions.  Additionally,
+      # resource B can define `action :create` and call `super()` to invoke A's
+      # action code.
+      #
+      # @param name [Symbol] The action name to define.
+      # @param recipe_block The recipe to run when the action is taken. This block
+      #   takes no parameters, and will be evaluated in a new context containing:
+      #
+      #   - The resource's public and protected methods (including attributes)
+      #   - The Chef Recipe DSL (file, etc.)
+      #   - super() referring to the parent version of the action (if any)
+      #
+      # @return The Action class implementing the action
+      #
+      def action(action, &recipe_block)
+        action = action.to_sym
+        create_action_provider_class.action(action, &recipe_block)
+        self.allowed_actions += [ action ]
+        default_action action if default_action == [ :nothing ]
+      end
+
+      #
+      # The list of allowed actions for the resource.
+      #
+      # The first time this is called with arguments, allowed_actions is replaced.
+      # Subsequent times will append.  Thus this:
+      #
+      # ```ruby
+      # allowed_actions :a
+      # allowed_actions :b, :c
+      # ```
+      #
+      # Is equivalent to this:
+      #
+      # ```ruby
+      # allowed_actions :a, :b, :c
+      # ```
+      #
+      # And when you inherit from a superclass, the replacement action is
+      # significant:
+      #
+      # ```ruby
+      # class A < Chef::Resource
+      #   allowed_actions :a, :b
+      # end
+      # class B < A
+      #   # This *replaces* allowed_actions, so allowed_actions is now :x, y
+      #   allowed_actions :x, :y
+      # end
+      # ```
+      #
+      # @param actions [Array<Symbol>] The list of actions to add/set.
+      #
+      # @return [Set<Symbol>] The list of actions, as symbols.
+      #
+      def allowed_actions(*actions)
+        if @allowed_actions
+          @allowed_actions |= actions
+        elsif actions.empty?
+          superclass.allowed_actions.dup
+        else
+          @allowed_actions = actions.to_set
+        end
+      end
+      def allowed_actions=(value)
+        @allowed_actions = value.to_set
+      end
+
+      #
+      # The action that will be run if no other action is specified.
+      #
+      # Defaults to the :nothing.
+      #
+      # @param action_name [Symbol,Array<Symbol>] The default action (or series
+      #   of actions) to use.
+      #
+      # @return [Symbol,Array<Symbol>] The default actions for the resource.
+      #
+      def default_action(action_name=NULL_ARG)
+        unless action_name.equal?(NULL_ARG)
```

Comment:
[Above](https://github.com/chef/chef/pull/3442/files#diff-135510b4e9e7ac268144bf6a7479a94dR891) (line 891) you use `!= NULL_ARG`, but here it's `#equal?`. Any reason?